src/circuit_proof/r1cs.rs

```diff
@@ -1,26 +1,31 @@
 #![allow(non_snake_case)]
+#![allow(non_camel_case_types)]
 
 use rand::{CryptoRng, Rng};
 
+use super::circuit::{Circuit, ProverInput, VerifierInput};
 use curve25519_dalek::ristretto::RistrettoPoint;
 use curve25519_dalek::scalar::Scalar;
+use errors::R1CSError;
 use generators::PedersenGenerators;
-use std::iter::FromIterator;
 
-// use circuit::{Circuit, ProverInput, VerifierInput};
-use super::circuit::{Circuit, ProverInput, VerifierInput};
+#[derive(Clone, Debug)]
+pub enum Variable {
+    Committed(usize),        // high-level variable
+    MultiplierLeft(usize),   // low-level variable, left input of multiplication gate
+    MultiplierRight(usize),  // low-level variable, right input of multiplication gate
+    MultiplierOutput(usize), // low-level variable, output multiplication gate
+}
 
-// This is a stripped-down version of the Bellman r1cs representation, for the purposes of
-// learning / understanding. The eventual goal is to write this as a BulletproofsConstraintSystem
-// that implements the Bellman ConstraintSystem trait, so we can use that code/logic.
-// (That would require the bellman code to be decoupled from the underlying pairings.)
+pub type Assignment = Result<Scalar, R1CSError>;
 
-/// Represents a variable in our constraint system, where the value represents the index.
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-pub struct Variable(usize);
+pub fn err_assignment() -> Assignment {
+    Err(R1CSError::InvalidVariableAssignment)
+}
 
 /// Represents a linear combination of some variables multiplied with their scalar coefficients,
-/// plus a scalar. E.g. LC = variable[0]*scalar[0] + variable[1]*scalar[1] + scalar
+/// plus a scalar. The linear combination is supposed to evaluate to zero.
+/// E.g. LC = 0 = variable[0]*scalar[0] + variable[1]*scalar[1] + scalar
```

Comment:
Is evaluating to zero really a property of the `LinearCombination`, or of the way that the `ConstraintSystem` will use the linear combinations?

Reply:
The latter - I will clarify that in the comment.